### PR TITLE
Spack environment for ATDM trilinos

### PR DESCRIPTION
**Pull request by fryeguy52**
Spack environment for ATDM trilinos

### CODE REVIEW

File: cmake/std/atdm/spack-rhel6/environment.sh [hunk 1]
```diff
@@ -131,11 +131,13 @@ export ATDM_CONFIG_NETCDF_LIBS="-L${BOOST_ROOT}/lib;-L${NETCDF_ROOT}/lib;-L${NET
...
 # use them in a shared lib build :-(
 
-export ATDM_CONFIG_BINUTILS_LIBS="${BINUTILS_ROOT}/lib/libbfd.so;${BINUTILS_ROOT}/lib64/libiberty.a"
+export ATDM_CONFIG_BINUTILS_LIBS="${BINUTILS_ROOT}/lib/libbfd.so;${BINUTILS_ROOT}/lib64/libiberty.a;${GETTEXT_ROOT}/lib/libintl.a;${LIBICONV_ROOT}/lib/libiconv.so"
 # NOTE: Above, we have to explicitly set the libs to use libbdf.so instead of
 # libbdf.a because the former works and the latter does not and TriBITS is set
 # up to only find static libs by default!
 unset BINUTILS_ROOT
+# NOTE: Above, you have to unset this or the configure dues in in the compiler
+# check.  What the heck?
```
Comment (mhoemmen) [inline cmake/std/atdm/spack-rhel6/environment.sh]: Did you mean to write this?

Comment (mhoemmen) [reply cmake/std/atdm/spack-rhel6/environment.sh]: oops, I didn't mean to hit approve; just cancelled

Comment (bartlettroscoe) [reply cmake/std/atdm/spack-rhel6/environment.sh]: Just pushed commit 5410b315ba738ea071ce754e31b11b1a873b5e4f that hopefully improves this comment.  Please let me know.  Thanks!

File: cmake/std/atdm/spack-rhel6/environment.sh [hunk 2]
```diff
@@ -136,8 +136,8 @@ export ATDM_CONFIG_BINUTILS_LIBS="${BINUTILS_ROOT}/lib/libbfd.so;${BINUTILS_ROOT
 # libbdf.a because the former works and the latter does not and TriBITS is set
 # up to only find static libs by default!
 unset BINUTILS_ROOT
-# NOTE: Above, you have to unset this or the configure dues in in the compiler
-# check.  What the heck?
+# NOTE: Above, you have to unset BINUTILS_ROOT or the CMake configure of
+# Trilinos dies in the compiler check.  What the heck?
```
Comment (mhoemmen) [inline cmake/std/atdm/spack-rhel6/environment.sh]: I'm objecting to "What the heck?".  While not obscene, it conveys an impression to users that we don't know what we're doing.  That impression may be not far from the truth ;-) but it's unnecessarily unsettling.

Comment (bartlettroscoe) [reply cmake/std/atdm/spack-rhel6/environment.sh]: @mhoemmen said:

> I'm objecting to "What the heck?".  While not obscene, it conveys an impression to users that we don't know what we're doing. That impression may be not far from the truth ;-) but it's unnecessarily unsettling.

This is scary that setting BINUTILS_ROOT breaks the compiler check for CMake.  We would have to track that down but this is scary.

Comment (mhoemmen) [reply cmake/std/atdm/spack-rhel6/environment.sh]: The NOTE already conveys that idea.

Look, people use this software to simulate Serious Things.  We need to sound serious here.  If something is broken, file an issue and put the issue number in the comment.

**Comment by bartlettroscoe** [comment]
NOTE: I just did a reset/cherry-pick operation to rebase the unique commits on this back to a version of Trilinos 'develop' on/before 2019-05-15.  This is so that we can carefully merge specific versions of Trilinos 'develop' into this branch locally so that we can carefully test against SPARC and EMPIRE with know working versions.  See details below.

<details>
<summary><b>DETAILS</b> (click to expand)</summary>

.

Okay, so I reset the branch 'spack-rhel6-env' back to the commit:

```
e6f6db4 "Merge remote branch 'origin/develop' into spack-rhel6-env"
Author: Joe Frye <jfrye@sandia.gov>
Date:   Wed May 15 12:57:31 2019 -0600 (3 weeks ago)
```

and then cherry-picked the unique commits form the branch 'sparc-rhel6-env' back on to the branch.  This ensrues that we base this off of a version of Trilinos before the last update of Triilnos by EMPIRE from 2019-05-30.  That is critical for testing against EMPIRE and SPARC.

To keep a record of the exact state of the branch 'spack-rhel6-env' before this reset and cherry-picking, I pushed to:

```
$ git push rab-github spack-rhel6-env-tmp:spack-rhel6-env-20190605
Counting objects: 5646, done.
Delta compression using up to 32 threads.
Compressing objects: 100% (1027/1027), done.
Writing objects: 100% (3957/3957), 1.70 MiB | 1.74 MiB/s, done.
Total 3957 (delta 3217), reused 3549 (delta 2894)
remote: Resolving deltas: 100% (3217/3217), completed with 972 local objects.
remote: 
remote: Create a pull request for 'spack-rhel6-env-20190605' on GitHub by visiting:
remote:      https://github.com/bartlettroscoe/Trilinos/pull/new/spack-rhel6-env-20190605
remote: 
To git@github.com:bartlettroscoe/Trilinos.git
 * [new branch]      spack-rhel6-env-tmp -> spack-rhel6-env-20190605
```

To see the unique commits on the branch before and after this reset/cherry-pick we look at:

```
$ git log-oneline --no-merges rab-github/spack-rhel6-env-20190605 --not trilinos/develop 
84e0ec1 "Update to match what is in EMPIRE BuildScripts/ALL/configure-trilinos.sh" <rabartl@sandia.gov> [Wed Jun 5 10:54:03 2019 -0600] (5 hours ago)
5410b31 "Fix a comment about unsetting BINUTILS_ROOT" <rabartl@sandia.gov> [Wed Jun 5 10:18:03 2019 -0600] (6 hours ago)
3bd163f "Make binutils work, clean up formatting" <rabartl@sandia.gov> [Tue Jun 4 17:34:05 2019 -0600] (22 hours ago)
ecd3c8e "Load gcc@7.2.0 built with native compiler, comment out debug echos" <rabartl@sandia.gov> [Tue Jun 4 16:32:39 2019 -0600] (23 hours ago)
31c3557 "Switch from hdf5-1.8.20 to 1.8.21" <rabartl@sandia.gov> [Tue Jun 4 11:57:58 2019 -0600] (28 hours ago)
8f3cf3b "squish: change hdf5 module version" <jfrye@sandia.gov> [Tue May 7 13:41:03 2019 -0600] (4 weeks ago)
312afcf "SQUISH" <jfrye@sandia.gov> [Mon May 6 08:46:40 2019 -0600] (4 weeks ago)
e2b9f19 "iterating on spack-rhel6 env" <jfrye@sandia.gov> [Tue Apr 30 12:05:55 2019 -0600] (5 weeks ago)
c438a94 "new spack rhel6 env" <jfrye@sandia.gov> [Thu Mar 28 15:39:54 2019 -0600] (5 weeks ago)
deb21fe "spack rhel6 env script" <jfrye@sandia.gov> [Wed Mar 20 13:31:39 2019 -0600] (5 weeks ago)
1563406 "remove intel compiler from spack rhel6 atdm env script" <jfrye@sandia.gov> [Thu Mar 28 15:44:55 2019 -0600] (5 weeks ago)
7968be3 "new spack rhel6 env" <jfrye@sandia.gov> [Thu Mar 28 15:39:54 2019 -0600] (5 weeks ago)
c480f70 "spack rhel6 env script" <jfrye@sandia.gov> [Wed Mar 20 13:31:39 2019 -0600] (5 weeks ago)
```

and compare to:

```
$ git log-oneline  HEAD --not trilinos/develop
75a1a94 "Update to match what is in EMPIRE BuildScripts/ALL/configure-trilinos.sh" <rabartl@sandia.gov> [Wed Jun 5 10:54:03 2019 -0600] (10 minutes ago)
1f4e907 "Fix a comment about unsetting BINUTILS_ROOT" <rabartl@sandia.gov> [Wed Jun 5 10:18:03 2019 -0600] (10 minutes ago)
a22b8ac "Make binutils work, clean up formatting" <rabartl@sandia.gov> [Tue Jun 4 17:34:05 2019 -0600] (10 minutes ago)
77b0e1f "Load gcc@7.2.0 built with native compiler, comment out debug echos" <rabartl@sandia.gov> [Tue Jun 4 16:32:39 2019 -0600] (11 minutes ago)
ddb6451 "Switch from hdf5-1.8.20 to 1.8.21" <rabartl@sandia.gov> [Tue Jun 4 11:57:58 2019 -0600] (11 minutes ago)
e6f6db4 "Merge remote branch 'origin/develop' into spack-rhel6-env" <jfrye@sandia.gov> [Wed May 15 12:57:31 2019 -0600] (3 weeks ago)
8f3cf3b "squish: change hdf5 module version" <jfrye@sandia.gov> [Tue May 7 13:41:03 2019 -0600] (4 weeks ago)
312afcf "SQUISH" <jfrye@sandia.gov> [Mon May 6 08:46:40 2019 -0600] (4 weeks ago)
e2b9f19 "iterating on spack-rhel6 env" <jfrye@sandia.gov> [Tue Apr 30 12:05:55 2019 -0600] (5 weeks ago)
c438a94 "new spack rhel6 env" <jfrye@sandia.gov> [Thu Mar 28 15:39:54 2019 -0600] (5 weeks ago)
deb21fe "spack rhel6 env script" <jfrye@sandia.gov> [Wed Mar 20 13:31:39 2019 -0600] (5 weeks ago)
1563406 "remove intel compiler from spack rhel6 atdm env script" <jfrye@sandia.gov> [Thu Mar 28 15:44:55 2019 -0600] (5 weeks ago)
7968be3 "new spack rhel6 env" <jfrye@sandia.gov> [Thu Mar 28 15:39:54 2019 -0600] (5 weeks ago)
c480f70 "spack rhel6 env script" <jfrye@sandia.gov> [Wed Mar 20 13:31:39 2019 -0600] (5 weeks ago)
```

So we see the commits up to:

```
8f3cf3b "squish: change hdf5 module version" <jfrye@sandia.gov> [Tue May 7 13:41:03 2019 -0600] (4 weeks ago)
```

are identical and have the same SHA1s.  It is only after the merge commit:

```
e6f6db4 "Merge remote branch 'origin/develop' into spack-rhel6-env" <jfrye@sandia.gov> [Wed May 15 12:57:31 2019 -0600] (3 weeks ago)
```

that we see new SHA1s in the last 5 commits:

```
84e0ec1 "Update to match what is in EMPIRE BuildScripts/ALL/configure-trilinos.sh" <rabartl@sandia.gov> [Wed Jun 5 10:54:03 2019 -0600] (5 hours ago)
5410b31 "Fix a comment about unsetting BINUTILS_ROOT" <rabartl@sandia.gov> [Wed Jun 5 10:18:03 2019 -0600] (6 hours ago)
3bd163f "Make binutils work, clean up formatting" <rabartl@sandia.gov> [Tue Jun 4 17:34:05 2019 -0600] (22 hours ago)
ecd3c8e "Load gcc@7.2.0 built with native compiler, comment out debug echos" <rabartl@sandia.gov> [Tue Jun 4 16:32:39 2019 -0600] (23 hours ago)
31c3557 "Switch from hdf5-1.8.20 to 1.8.21" <rabartl@sandia.gov> [Tue Jun 4 11:57:58 2019 -0600] (28 hours ago)
```

Now, the last check is to diff the Trilinos/cmake/std/atdm/ directories and see if there are any changes:

```
$ git diff --name-status HEAD --not rab-github/spack-rhel6-env-20190605 -- cmake/std/atdm/
[empty]
```

So that is the proof that this was a correct manupluation of this branch.

</details>

File: cmake/std/atdm/spack-rhel6/environment.sh (hunk 2 above)
Comment (bartlettroscoe) [reply cmake/std/atdm/spack-rhel6/environment.sh]: Fixed in commit 15dce2b

**Comment by bartlettroscoe** [comment]
FYI: I merged (fast-forwarded) the branch 'spack-rhel6-env-for-empire' to her since it worked for EMPIRE and and the version of Trilinos from May 30 merged in also seems to work for SPARC (see [CDOFA-43](https://sems-atlassian-srn.sandia.gov/browse/CDOFA-43)).

File: cmake/std/atdm/spack-rhel6/environment.sh (hunk 2 above)
Comment (mhoemmen) [reply cmake/std/atdm/spack-rhel6/environment.sh]: Argh, I realized that @bartlettroscoe had written this.  He is allowed to say "what the heck?!?" whenever he wants :-D

File: cmake/std/atdm/utils/get_known_system_name.sh [hunk 3]
```diff
@@ -65,6 +65,9 @@ elif [[ $ATDM_CONFIG_REAL_HOSTNAME == "mutrino"* ]] ; then
 elif [[ $ATDM_CONFIG_REAL_HOSTNAME == "waterman"* ]] ; then
   ATDM_HOSTNAME=waterman
   ATDM_SYSTEM_NAME=waterman
+elif [[ $ATDM_CONFIG_BUILD_NAME == *"spack-rhel6"* ]] ; then
```
[the remark this answers is not in the document]
Comment (bartlettroscoe) [reply cmake/std/atdm/utils/get_known_system_name.sh]: You just need to make sure you load the spack env first.  Will provide instructions else where.

File: cmake/std/atdm/spack-rhel6/environment.sh [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    module load gcc-7.2.0/spack-netlib-lapack/3.8.0
+    export BLAS_ROOT=$NETLIB_LAPACK_ROOT
+    export LAPACK_ROOT=$NETLIB_LAPACK_ROOT
+
+#    export LAPACK_ROOT=/usr/lib64/atlas
+    export OMPI_CC=`which gcc`
+    export OMPI_FC=`which gfortran`
+    export ATDM_CONFIG_LAPACK_LIBS="-L${LAPACK_ROOT};-llapack"
+    export ATDM_CONFIG_BLAS_LIBS="-L${BLAS_ROOT}/lib;-lblas"
+elif [ "$ATDM_CONFIG_COMPILER" == "INTEL" ]; then
```
[the remark this answers is not in the document]
Comment (bartlettroscoe) [reply cmake/std/atdm/spack-rhel6/environment.sh]: I removed this block.

**Comment by bartlettroscoe** [comment]
Lots of testing with this but after the final tweaks, on my machine 'crf450', I ran first the Trilinos builds:

```
$ ./checkin-test-atdm-spack-rhel6.sh \
    spack-rhel-gnu-7.2.0-openmp-opt-dbg \
    sems-rhel6-gnu-7.2.0-openmp-opt-dbg \
    --enable-packages=Kokkos,Teuchos,SEACAS,Panzer \
    --make-options=-j20 --ctest-options=-j20 --local-do-all
```

which returned:

```
Thu Jun 13 16:17:55 MDT 2019

Enabled Packages: Kokkos, Teuchos, SEACAS, Panzer

Build test results:
-------------------
1) spack-rhel-gnu-7.2.0-openmp-opt-dbg => passed: passed=378,notpassed=0 (25.83 min)
2) sems-rhel6-gnu-7.2.0-openmp-opt-dbg => passed: passed=352,notpassed=0 (24.03 min)
```

<details>
<summary><b>DETAILED TEST RESULTS:</b> (click to expand)</summary>

.

```
1) spack-rhel-gnu-7.2.0-openmp-opt-dbg Results:
-----------------------------------------------

  passed: Trilinos/spack-rhel-gnu-7.2.0-openmp-opt-dbg: passed=378,notpassed=0
  
  Thu Jun 13 15:53:48 MDT 2019
  
  Enabled Packages: Kokkos, Teuchos, SEACAS, Panzer
  Hostname: crf450.srn.sandia.gov
  Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
  Build Dir: /home/rabartl/Trilinos.base/BUILDS/ATDM/SPACK-RHEL/CHECKIN/spack-rhel-gnu-7.2.0-openmp-opt-dbg
  
  CMake Cache Varibles: -GNinja -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=NIGHTLY -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=600.0 -GNinja -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_Kokkos:BOOL=ON -DTrilinos_ENABLE_Teuchos:BOOL=ON -DTrilinos_ENABLE_SEACAS:BOOL=ON -DTrilinos_ENABLE_Panzer:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=OFF
  Make Options: -j20
  CTest Options: -j20
  
  Pull: Not Performed
  Configure: Passed (1.20 min)
  Build: Passed (21.04 min)
  Test: Passed (3.59 min)
  
  100% tests passed, 0 tests failed out of 378
  
  Subproject Time Summary:
  Kokkos     =  78.46 sec*proc (27 tests)
  Panzer     = 3044.37 sec*proc (167 tests)
  SEACAS     = 462.69 sec*proc (46 tests)
  Teuchos    =  88.87 sec*proc (138 tests)
  
  Total Test time (real) = 215.20 sec
  
  Total time for spack-rhel-gnu-7.2.0-openmp-opt-dbg = 25.83 min


2) sems-rhel6-gnu-7.2.0-openmp-opt-dbg Results:
-----------------------------------------------

  passed: Trilinos/sems-rhel6-gnu-7.2.0-openmp-opt-dbg: passed=352,notpassed=0
  
  Thu Jun 13 16:17:52 MDT 2019
  
  Enabled Packages: Kokkos, Teuchos, SEACAS, Panzer
  Hostname: crf450.srn.sandia.gov
  Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
  Build Dir: /home/rabartl/Trilinos.base/BUILDS/ATDM/SPACK-RHEL/CHECKIN/sems-rhel6-gnu-7.2.0-openmp-opt-dbg
  
  CMake Cache Varibles: -GNinja -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=NIGHTLY -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=600.0 -GNinja -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_Kokkos:BOOL=ON -DTrilinos_ENABLE_Teuchos:BOOL=ON -DTrilinos_ENABLE_SEACAS:BOOL=ON -DTrilinos_ENABLE_Panzer:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=OFF
  Make Options: -j20
  CTest Options: -j20
  
  Pull: Not Performed
  Configure: Passed (0.34 min)
  Build: Passed (20.56 min)
  Test: Passed (3.14 min)
  
  100% tests passed, 0 tests failed out of 352
  
  Subproject Time Summary:
  Kokkos     = 117.32 sec*proc (27 tests)
  Panzer     = 2718.65 sec*proc (167 tests)
  SEACAS     =  33.01 sec*proc (20 tests)
  Teuchos    =  88.02 sec*proc (138 tests)
  
  Total Test time (real) = 187.97 sec
  
  Total time for sems-rhel6-gnu-7.2.0-openmp-opt-dbg = 24.03 min
```

</details>

I also tested this with SPARC and EMPIRE and those tests are documented in [CDOFA-43](https://sems-atlassian-srn.sandia.gov/browse/CDOFA-43).

Going to go ahead and merge this now.

**Pull request merged**